### PR TITLE
Improve All-Reduce fault-tolerance

### PR DESCRIPTION
**Pull request by justheuristic**
- [x] allow AllreduceRunner to tolerate clients that
   - [x] do not send some of their local tensors
   - [x] do not show up at at all after matchmaking is over
- [x] allow AllreduceRunner to tolerate full/aux peers that do not send some or all results
- [x] introduce timeout after which sender/reducer is considered failed
- [x] AllreduceRunner & DecentralizedAverager will no longer _send_error_to_peer
   - log spam is gone!
- report allreduce integrity
   - [x] TensorPartReducer will report the fraction of expected parts received if that fraction is not 1
   - [x] TensorPartContainer will report the fraction of parts that did not fail if that fraction is not 1
- miscellaneous improvements to Optimizer
  - [x] set good default sender/reducer timeouts
  - [x] pre-schedule state averaging ahead of time
  - [x] no longer block the entire peer if it is time to pre-schedule gradients but background state averaging is still underway

Test cases:
- [x] test with peers that fail early
- [x] test with peers that fail to send a certain part
- [x] test with peers that fail to reduce their part
- [x] test cancelling

Sanity checks:
- [x] run tests 100 times
- [x] benchmark_optimizer
- [x] test env 64+ nodes 4+ hours

### CODE REVIEW

File: hivemind/averaging/allreduce.py [hunk 1]
```diff
@@ -132,6 +148,8 @@ def should_delay_results(self, peer_id: PeerID) -> bool:
     async def run(self) -> AsyncIterator[torch.Tensor]:
         """Run all-reduce, return differences between averaged and original tensors as they are computed"""
         pending_tasks = set()
+        if self.sender_timeout is not None:
+            asyncio.create_task(self._handle_missing_senders())
```
Comment (borzunov) [inline hivemind/averaging/allreduce.py:152]: Such fire-and-forget calls always lead to "Task destroyed but is pending". Please fix that, e.g., save the task to a class field and `await` when finalizing.

File: hivemind/averaging/allreduce.py [hunk 2]
```diff
@@ -205,28 +262,32 @@ async def rpc_aggregate_part(
     ) -> AsyncIterator[averaging_pb2.AveragingData]:
         """a peer sends us a part of his tensor; we should average it with other peers and return the difference"""
         request: averaging_pb2.AveragingData = await anext(stream)
```
Comment (borzunov) [inline hivemind/averaging/allreduce.py:264]: Please handle `self.sender_timeout` for this first part as well.

File: hivemind/averaging/allreduce.py [hunk 3]
```diff
@@ -240,62 +301,62 @@ async def _accumulate_parts():
...
         else:
             error_code = averaging_pb2.MessageCode.Name(request.code)
-            logger.debug(f"{self} - peer {context.remote_id} sent {error_code}, allreduce cannot continue")
-            self.finalize(exception=AllreduceException(f"Peer {context.remote_id} sent {error_code}"))
+            logger.debug(f"{self} - peer {context.remote_id} sent {error_code}")
+            sender_index = self.sender_peer_ids.index(context.remote_id)
+            async with self.banlock:
+                if context.remote_id not in self.banned_senders:
+                    self.banned_senders.add(context.remote_id)
+                    self.tensor_part_reducer.on_sender_failed(sender_index)
```
Comment (borzunov) [inline hivemind/averaging/allreduce.py:317]: Please fix this code duplication (L305-308, L314-317, L352-355), this is likely to lead to a bug if someone decides to change the ban procedure (but changes only 1-2 of 3 copies).

**Review by borzunov — CHANGES_REQUESTED** [review]
Thanks for the PR! I've went through everything but tests and left my comments.

File: hivemind/utils/asyncio.py [hunk 4]
```diff
@@ -114,23 +114,30 @@ async def amap_in_executor(
...
+                raise future_or_exception
+            yield await future_or_exception
+            future_or_exception = await queue.get()
         await task
     finally:
-        if not task.done():
-            task.cancel()
+        task.cancel()
+        if task.done() and not task.cancelled():
+            task.exception()
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/utils/asyncio.py:137]: I specifically mean "if task did not send result or throw exception but we died anyway, silence the "task ... was never retrieved""

File: hivemind/utils/asyncio.py [hunk 5]
```diff
@@ -114,23 +114,30 @@ async def amap_in_executor(
...
-        future = await queue.get()
-        while future is not None:
-            yield await future
-            future = await queue.get()
+        future_or_exception = await queue.get()
+        while future_or_exception is not None:
+            if isinstance(future_or_exception, BaseException):
+                raise future_or_exception
+            yield await future_or_exception
+            future_or_exception = await queue.get()
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/utils/asyncio.py:132]: Previously, if input iterator threw an exception, this code would freeze forever

File: hivemind/optim/experimental/optimizer.py [hunk 6]
```diff
@@ -427,6 +432,8 @@ def _update_global_epoch(self, grad_scaler: Optional[GradScaler]) -> None:
 
             if self.use_gradient_averaging:
                 logger.log(self.status_loglevel, f"Beginning optimizer step #{self.local_epoch}")
+                self.state_averager.step(wait_for_delayed_updates=True)
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/optim/experimental/optimizer.py:435]: If previous optimizer step is still running, we wait to avoid modifying its gradients in-place while it may still need them

File: hivemind/averaging/partition.py [hunk 7]
```diff
@@ -209,23 +219,38 @@ async def accumulate_part(
...
+                self.reset_accumulators()
+
     def finalize(self):
         if not self.finished.is_set():
             if hasattr(self, "current_part_future"):
                 self.current_part_future.cancel()
                 del self.accumulator
             self.finished.set()
 
+            if self.num_parts != 0 and self.num_senders != 0:
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/averaging/partition.py:249]: it protects from zero division error

File: hivemind/averaging/allreduce.py (hunk 1 above)
Comment (justheuristic) [reply hivemind/averaging/allreduce.py]: GREAT catch: i've meant to add it to pending_tasks (and i just did so)

File: hivemind/averaging/allreduce.py (hunk 3 above)
Comment (justheuristic) [reply hivemind/averaging/allreduce.py]: replaced with a unified `await self._ban_sender(peer_id)` call

File: hivemind/averaging/partition.py [hunk 8]
```diff
@@ -91,21 +92,16 @@ def get_raw_input_parts(self, peer_index: int) -> Tuple[torch.Tensor, ...]:
...
 
     @torch.no_grad()
     async def iterate_input_parts_for(self, peer_index: int) -> AsyncIterator[runtime_pb2.Tensor]:
         """iterate serialized tensor parts for a peer at a given index. Run serialization in background."""
         assert not self._inputs_consumed_by_peer[peer_index], "input parts of a given peer are already deallocated."
         self._inputs_consumed_by_peer[peer_index] = True
-
-        async def _aiterate_parts():
-            for _ in range(self.num_parts_by_peer[peer_index]):
-                yield self._input_parts_by_peer[peer_index].popleft()
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/averaging/partition.py:105]: These are all links that reuse tensors' memory, we fixed it a couple of months ago

The additional memory is only in the metadata

File: hivemind/averaging/partition.py [hunk 9]
```diff
@@ -155,8 +160,6 @@ def finalize(self):
         if not self.finished.is_set():
             for peer_index in range(self.group_size):
                 self._inputs_consumed_by_peer[peer_index] = True
-                self._input_parts_by_peer[peer_index].clear()
-                self._output_parts_by_peer[peer_index].clear()
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/averaging/partition.py:159]: done. will introduce 16 such peers in test env and test if they're fine

File: hivemind/averaging/partition.py [hunk 10]
```diff
@@ -178,18 +181,24 @@ def __init__(self, part_shapes: Sequence[torch.Size], num_senders: int):
         self.denominator = 0.0  # total weight accumulated from all peers for current part
         self.current_part_future = asyncio.Future()
         self.finished = asyncio.Event()
+
+        self.num_parts_received = [0 for _ in range(self.num_senders)]
+        self.sender_failed_after = [float("inf") for _ in range(self.num_senders)]
+        self.current_senders = self.num_senders
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/averaging/partition.py]: done

File: hivemind/averaging/allreduce.py (hunk 2 above)
Comment (justheuristic) [reply hivemind/averaging/allreduce.py]: done

File: hivemind/averaging/partition.py [hunk 11]
```diff
@@ -209,23 +219,38 @@ async def accumulate_part(
...
         return await current_part_future
 
+    def on_sender_failed(self, sender_index: int):
+        """Exclude that sender's data for averaging any parts that it did not submit yet."""
+        self.sender_failed_after[sender_index] = self.num_parts_received[sender_index]
+        if self.current_part_index == self.num_parts_received[sender_index]:
+            self.current_senders -= 1
+            if self.current_part_accumulated_from == self.current_senders:
+                self.current_part_future.set_result(self.accumulator.div_(self.denominator))
+                self.reset_accumulators()
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/averaging/partition.py]: done

File: hivemind/averaging/partition.py [hunk 12]
```diff
@@ -178,18 +183,25 @@ def __init__(self, part_shapes: Sequence[torch.Size], num_senders: int):
         self.denominator = 0.0  # total weight accumulated from all peers for current part
         self.current_part_future = asyncio.Future()
         self.finished = asyncio.Event()
+
+        self.num_parts_received = [0 for _ in range(self.num_senders)]
+        self.sender_failed_after = [float("inf") for _ in range(self.num_senders)]
+        self.\
+            num_current_senders = self.num_senders
```
Comment (borzunov) [inline hivemind/averaging/partition.py]:
```suggestion
        self.num_current_senders = self.num_senders
```

Comment (justheuristic) [reply hivemind/averaging/partition.py]: thx

File: hivemind/utils/asyncio.py [hunk 13]
```diff
@@ -114,23 +114,30 @@ async def amap_in_executor(
...
     async def _put_items():
-        async for args in azip(*iterables):
-            await queue.put(loop.run_in_executor(executor, func, *args))
-        await queue.put(None)
+        try:
+            async for args in azip(*iterables):
+                await queue.put(loop.run_in_executor(executor, func, *args))
+            await queue.put(None)
+        except BaseException as e:
+            await queue.put(e)  # note: there is no chance that iterables
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/utils/asyncio.py:122]: fixd

File: hivemind/utils/asyncio.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/utils/asyncio.py:137]: Let's remove L136-137 because `.cancel()` already suppresses this message.

Indeed, the message is printed only when `task.__log_traceback == True` ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L94)) and `task.cancel()` sets it to `False` ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L151)), exactly as `task.exception()` does ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L217)).

File: hivemind/utils/asyncio.py [hunk 14]
```diff
@@ -114,23 +114,30 @@ async def amap_in_executor(
...
     async def _put_items():
-        async for args in azip(*iterables):
-            await queue.put(loop.run_in_executor(executor, func, *args))
-        await queue.put(None)
+        try:
+            async for args in azip(*iterables):
+                await queue.put(loop.run_in_executor(executor, func, *args))
+            await queue.put(None)
+        except BaseException as e:
+            await queue.put(e)
```
Comment (borzunov) [inline hivemind/utils/asyncio.py:122]: Suggestion: let's do it like this and revert the change below. This will remove this edge case as well as unnecessary `isinstance()` check.

```suggestion
            future = asyncio.Future()
            future.set_exception(e)
            await queue.put(future)
```

File: tests/test_allreduce_fault_tolerance.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+from hivemind.proto import averaging_pb2
+from hivemind.utils.asyncio import as_aiter, azip, enter_asynchronously
+from hivemind.utils.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class Fault(Enum):
+    NONE = auto()
+    FAIL_BEFORE = auto()
```
Comment (borzunov) [inline tests/test_allreduce_fault_tolerance.py:25]:
```suggestion
    FAIL_AWOL = auto()
```

[Cambridge Dictionary](https://dictionary.cambridge.org/dictionary/english/awol)

File: tests/test_allreduce_fault_tolerance.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+
+class FaultyAllReduceRunner(AllReduceRunner):
+    def __init__(self, *args, fault: Fault, **kwargs):
+        self.fault = fault
+        super().__init__(*args, **kwargs)
+
+    async def rpc_aggregate_part(self, stream, context) -> AsyncIterator[averaging_pb2.AveragingData]:
+        if self.fault == Fault.FAIL_REDUCING:
+            yield averaging_pb2.AveragingData(code=averaging_pb2.INTERNAL_ERROR)
```
Comment (borzunov) [inline tests/test_allreduce_fault_tolerance.py:91]: Can we also test the case when `INTERNAL_ERROR` and `CANCEL` are the 2nd message? Should be useful since the 1st and all other messages are processed separately.

File: tests/test_allreduce_fault_tolerance.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    [
+        (Fault.NONE, Fault.FAIL_BEFORE),
+        (Fault.FAIL_BEFORE, Fault.FAIL_BEFORE),
+        (Fault.FAIL_SENDING, Fault.FAIL_SENDING),
+        (Fault.FAIL_SENDING, Fault.FAIL_BEFORE),
+        (Fault.FAIL_SENDING, Fault.FAIL_REDUCING),
+        (Fault.NONE, Fault.CANCEL),
+    ],
+)
+def test_fault_tolerance(fault0: Fault.NONE, fault1: Fault.NONE):
```
Comment (borzunov) [inline tests/test_allreduce_fault_tolerance.py:128]:
```suggestion
def test_fault_tolerance(fault0: Fault, fault1: Fault):
```

File: tests/test_allreduce_fault_tolerance.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+
+    dht = hivemind.DHT(start=True)
+
+    averagers = []
+    for i in range(5):
+        averager = FaultyAverager(
+            _make_tensors(),
+            hivemind.DHT(initial_peers=dht.get_visible_maddrs(), start=True),
+            prefix="test",
+            request_timeout=0.3,
```
Comment (borzunov) [inline tests/test_allreduce_fault_tolerance.py:140]: Can we also test timeouts?

File: tests/test_allreduce_fault_tolerance.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+        if all(fault == Fault.FAIL_SENDING for fault in (fault0, fault1)):
+            assert fault0 != Fault.FAIL_REDUCING and fault1 != Fault.FAIL_REDUCING
+            assert abs(diff[: len(diff) // 2]).max() < 1e-5
+        elif any(fault == Fault.CANCEL for fault in (fault0, fault1)):
+            pass  # late cancel may result in an arbitrary mix of averaging results with and without the cancelled peer
+        elif fault0 == Fault.NONE:  # only peer1 in client mode may have failed
+            assert abs(diff).max() < 1e-5
+        else:
+            assert (abs(diff) < 1e-5).numpy().mean() > 0.5
```
Comment (borzunov) [inline tests/test_allreduce_fault_tolerance.py:185]: Can we test the expected behavior on `FAIL_REDUCING`? It would be to be ensure that the averager uses a correct local part in this case (not some another part that will break the model completely).

Comment (justheuristic) [reply tests/test_allreduce_fault_tolerance.py:192]: added that case

File: hivemind/averaging/allreduce.py [hunk 20]
```diff
@@ -168,25 +195,51 @@ async def _communicate_with_peer(self, peer_id: PeerID):
...
+
+                part_index = 0
+
+                def _try_deserialize(msg):
+                    if msg.code != averaging_pb2.AVERAGED_PART:
+                        return AllreduceException(f"{peer_id} sent {averaging_pb2.MessageCode.Name(msg.code)}"), msg
+                    try:
+                        return deserialize_torch_tensor(msg.tensor_part), msg
+                    except Exception as e:
+                        return e, msg
```
Comment (borzunov) [inline hivemind/averaging/allreduce.py:215]:
```suggestion
                    if msg.code != averaging_pb2.AVERAGED_PART:
                        raise AllreduceException(f"{peer_id} sent {averaging_pb2.MessageCode.Name(msg.code)}")
                    return deserialize_torch_tensor(msg.tensor_part), msg
```

Can we do that? Since `amap_in_executor()` is tolerant to exceptions

**Review by borzunov — APPROVED** [review]

File: tests/test_allreduce_fault_tolerance.py (hunk 18 above)
Comment (justheuristic) [reply tests/test_allreduce_fault_tolerance.py:154]: [added two test conditions: timeout while sending and while reducing]
